File: sdk/npm/src/providers/mailcat-ai.ts
import { InternalEmailInfo, Email, Channel } from "../types";
import { normalizeEmail } from "../normalize";
import { fetchWithTimeout } from "../retry";

const CHANNEL: Channel = "mailcat-ai";
const API_BASE = "https://api.mailcat.ai";

interface CreateMailboxResponse {
  data?: {
    email?: string;
    token?: string;
  };
}

interface MailcatMessage {
  id?: string;
  from?: string;
  sender?: string;
  to?: string;
  subject?: string;
  text?: string;
  html?: string;
  date?: string;
  receivedAt?: string;
  isRead?: boolean;
}

interface InboxResponse {
  data?: MailcatMessage[];
  meta?: {
    mailbox?: string;
    unread?: number;
    pagination?: {
      offset?: number;
      limit?: number;
      totalCount?: number;
      hasMore?: boolean;
    };
  };
}

/**
 * 创建 mailcat.ai 临时邮箱
 * API: POST /mailboxes（无 body）
 * 返回 token 用于后续获取邮件
 */
export async function generateEmail(): Promise<InternalEmailInfo> {
  const response = await fetchWithTimeout(`${API_BASE}/mailboxes`, {
    method: "POST",
    headers: { Accept: "application/json" },
  });
  if (!response.ok) {
    throw new Error(`mailcat-ai: 创建邮箱失败 ${response.status}`);
  }

  const resp = (await response.json()) as CreateMailboxResponse;
  const address = String(resp.data?.email || "").trim();
  const token = String(resp.data?.token || "").trim();

  if (!address || !address.includes("@")) {
    throw new Error("mailcat-ai: 创建邮箱响应缺少有效地址");
  }
  if (!token) {
    throw new Error("mailcat-ai: 创建邮箱响应缺少 token");
  }

  return {
    channel: CHANNEL,
    email: address,
    token,
  };
}

/**
 * 获取 mailcat.ai 邮件列表
 * API: GET /inbox，需要 Authorization: Bearer {token}
 * @param token - Bearer token（创建邮箱时返回）
 * @param email - 邮箱地址
 */
export async function getEmails(
  token: string,
  email: string,
): Promise<Email[]> {
  if (!token) {
    throw new Error("mailcat-ai: token 不能为空");
  }

  const address = String(email || "").trim();
  if (!address) {
    throw new Error("mailcat-ai: 邮箱地址为空");
  }

  const response = await fetchWithTimeout(`${API_BASE}/inbox`, {
    method: "GET",
    headers: {
      Accept: "application/json",
      Authorization: `Bearer ${token}`,
    },
  });
  if (!response.ok) {
    throw new Error(`mailcat-ai: 获取邮件失败 ${response.status}`);
  }

  const resp = (await response.json()) as InboxResponse;
  const messages: MailcatMessage[] = Array.isArray(resp.data) ? resp.data : [];

  return messages.map((msg) => {
    const raw: Record<string, unknown> = {
      id: msg.id || "",
      from: msg.from || msg.sender || "",
      to: msg.to || address,
      subject: msg.subject || "",
      text: msg.text || "",
      html: msg.html || "",
      date: msg.date || msg.receivedAt || "",
      isRead: msg.isRead || false,
    };
    return normalizeEmail(raw, address);
  });
}

File: sdk/rust/src/providers/mailcat_ai.rs
/*!
 * mailcat.ai — https://mailcat.ai
 * 流程：POST /mailboxes 创建邮箱（返回 Bearer token）
 * 收信：GET /inbox 带 Authorization: Bearer {token}
 */

use crate::config::{block_on, http_client};
use crate::normalize::normalize_email;
use crate::types::{Channel, Email, EmailInfo};
use serde_json::{json, Value};

const API_BASE: &str = "https://api.mailcat.ai";

/// 创建 mailcat.ai 临时邮箱
pub fn generate_email() -> Result<EmailInfo, String> {
    block_on(async {
        let client = http_client();

        let resp = client
            .post(format!("{}/mailboxes", API_BASE))
            .send()
            .await
            .map_err(|e| format!("mailcat-ai: 创建邮箱失败: {}", e))?;

        if !resp.status().is_success() {
            return Err(format!(
                "mailcat-ai: 创建邮箱失败 HTTP {}",
                resp.status()
            ));
        }

        let body: Value = resp
            .json()
            .await
            .map_err(|e| format!("mailcat-ai: 解析响应失败: {}", e))?;

        let email = body["data"]["email"]
            .as_str()
            .ok_or("mailcat-ai: 响应中未找到邮箱地址")?
            .to_string();

        let token = body["data"]["token"]
            .as_str()
            .ok_or("mailcat-ai: 响应中未找到 token")?
            .to_string();

        Ok(EmailInfo {
            channel: Channel::MailcatAi,
            email,
            token: Some(token),
            expires_at: None,
            created_at: None,
        })
    })
}

/// 获取 mailcat.ai 邮件列表
pub fn get_emails(token: &str, email: &str) -> Result<Vec<Email>, String> {
    let address = email.trim();
    if address.is_empty() {
        return Err("mailcat-ai: 邮箱地址为空".into());
    }

    block_on(async {
        let client = http_client();

        let resp = client
            .get(format!("{}/inbox", API_BASE))
            .header("Authorization", format!("Bearer {}", token))
            .send()
            .await
            .map_err(|e| format!("mailcat-ai: 获取邮件失败: {}", e))?;

        if !resp.status().is_success() {
            return Err(format!(
                "mailcat-ai: 获取邮件失败 HTTP {}",
                resp.status()
            ));
        }

        let body: Value = resp
            .json()
            .await
            .map_err(|e| format!("mailcat-ai: 解析响应失败: {}", e))?;

        let messages = match body["data"].as_array() {
            Some(arr) => arr.clone(),
            None => return Ok(Vec::new()),
        };

        if messages.is_empty() {
            return Ok(Vec::new());
        }

        let mut out = Vec::with_capacity(messages.len());
        for msg in &messages {
            let raw = json!({
                "id": msg.get("id"),
                "from": msg.get("from").or_else(|| msg.get("sender")),
                "to": address,
                "subject": msg.get("subject"),
                "date": msg.get("date").or_else(|| msg.get("receivedAt")),
                "html": msg.get("html").or_else(|| msg.get("body")).and_then(|v| v.as_str()).unwrap_or(""),
                "text": msg.get("text").and_then(|v| v.as_str()).unwrap_or(""),
            });
            out.push(normalize_email(&raw, address));
        }
        Ok(out)
    })
}
